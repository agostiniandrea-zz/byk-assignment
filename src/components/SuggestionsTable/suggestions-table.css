.suggestions-table {
  position: relative;
  width: 100%;
  height: auto;
}