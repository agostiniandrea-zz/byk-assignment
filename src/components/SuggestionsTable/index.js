import React from 'react';
import './suggestions-table.css';
import SuggestionItem from '../SuggestionItem';

class SuggestionsTable extends React.Component {
  render() {
    return (
      <div className="suggestions-table">
        {this.props.items && this.props.items.map((object, i) =>
          <SuggestionItem nrResults={object.nrResults} searchterm={object.searchterm} typedString={this.props.typedString} key={i} />
        )}
      </div>
    );
  }
};

export default SuggestionsTable;