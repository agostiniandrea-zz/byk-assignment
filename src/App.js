import React, { Component } from 'react';

import SuggestionsTable from './components/SuggestionsTable';

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      error: null,
      searchValue: '',
      showSuggestions: false,
      suggestions: []
    };

    this.onBlur = this.onBlur.bind(this);
    this.cleanSearch = this.cleanSearch.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleOnKeyUp = this.handleOnKeyUp.bind(this);
    this.handleSuggestionsVisibility = this.handleSuggestionsVisibility.bind(this);
    this.searchInfo = this.searchInfo.bind(this);

    this.textInput = React.createRef();
  }

  onBlur() {
    this.deleteSuggestions();
  }

  deleteSuggestions() {
    this.setState({
      suggestions: []
    });
  }

  handleChange(event) {
    this.handleSuggestionsVisibility();
    this.setState({ searchValue: event.target.value });
  }

  handleOnKeyUp() {
    this.handleSuggestionsVisibility(true);
    if (this.state.searchValue) {
      this.searchInfo();
    } else {
      this.deleteSuggestions();
    }
  }

  handleSuggestionsVisibility(value = false) {
    this.setState({ showSuggestions: value });
  }

  searchInfo() {
    fetch(`/search?q=${this.state.searchValue}`)
      .then(res => res.json())
      .then((result) => {
        this.setState({ suggestions: result.suggestions });
      },
        (error) => {
          this.setState({
            error
          });
        }
      );
  }

  cleanSearch() {
    this.setState({
      searchValue: '',
      suggestions: []
    });

    if (this.textInput && this.textInput.focus) {
      this.textInput.current.focus();
    }
  }

  render() {
    return (
      <div className="app">
        <div className="search-bar">
          {this.state.searchValue && <label className="delete-label" onClick={this.cleanSearch}>X</label>}
          <img alt="" className="search-icon" src={require('./assets/icons/search-icon.png')} />
          <input
            className="input-field"
            placeholder="Zoeken"
            maxLength="40"
            type="text"
            ref={this.textInput}
            value={this.state.searchValue}
            onBlur={this.onBlur}
            onChange={this.handleChange}
            onKeyUp={this.handleOnKeyUp}
          />
        </div>
        {this.state.showSuggestions && this.state.suggestions && <SuggestionsTable items={this.state.suggestions} typedString={this.state.searchValue} />}
      </div>
    );
  }
}

export default App;
