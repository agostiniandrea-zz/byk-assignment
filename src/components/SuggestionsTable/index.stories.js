import React from 'react';
import SuggestionsTable from './index';

export default { title: 'SuggestionsTable' };

const mockData = require("../../../api/_search.get.json");
const suggestions = mockData.suggestions;

export const withoutItems = () => <SuggestionsTable />;
export const withItems = () => <SuggestionsTable items={suggestions} typedString="trui" />;