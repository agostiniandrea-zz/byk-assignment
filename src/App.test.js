import React from 'react';
import { shallow } from 'enzyme';
import App from './App';
import SuggestionsTable from './components/SuggestionsTable';

const typedSearch = "trui";

let wrapper;

beforeEach(() => {
  wrapper = shallow(<App />);
})

afterEach(() => {
  wrapper.unmount();
});

describe('App.js', () => {
  it('renders without crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('updates properly state.searchValue - DOM way', () => {
    wrapper.instance().setState({ searchValue: typedSearch });
    expect(wrapper.find(".delete-label").exists()).toBe(true);
    wrapper.find(".delete-label").simulate("click");
    expect(wrapper.instance().state.searchValue).toBe("");
    expect(wrapper.instance().state.suggestions).toStrictEqual([]);
  });

  it('updates properly state.searchValue - manual way', () => {
    wrapper.instance().setState({ searchValue: typedSearch });
    expect(wrapper.find(".delete-label").exists()).toBe(true);
    wrapper.instance().setState({ searchValue: "" });
    expect(wrapper.instance().state.searchValue).toBe("");
    expect(wrapper.instance().state.suggestions).toStrictEqual([]);
  });

  it('toggles state.showSuggestions', () => {
    wrapper.find('.input-field').simulate("change", { target: { value: typedSearch } })
    wrapper.find(".input-field").simulate("keyDown");
    expect(wrapper.instance().state.showSuggestions).toBe(false);
    wrapper.find(".input-field").simulate("keyUp");
    expect(wrapper.instance().state.showSuggestions).toBe(true);
  });

  it('renders SuggestionsTable component', () => {
    expect(wrapper.find(SuggestionsTable).length).toBe(wrapper.instance().state.suggestions.length);
  });
});