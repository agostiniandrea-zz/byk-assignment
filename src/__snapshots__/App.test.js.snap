// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.js renders without crashing 1`] = `ShallowWrapper {}`;
